### PR TITLE
Incorporated the Idempotency in Kafka Input Operator

### PR DESCRIPTION
**Pull request by chaithu14**

### CODE REVIEW

**Comment by chandnisingh** [comment]
There weren't any tests added for idempotency. 
Since you added tests for Kinesis  is there  a problem to add tests for Kafka?

**Comment by chaithu14** [comment]
Added unit test case and updated the code.

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractKafkaInputOperator.java [hunk 1]
```diff
@@ -98,6 +135,14 @@ public void setup(OperatorContext context)
   {
     logger.debug("consumer {} topic {} cacheSize {}", consumer, consumer.getTopic(), consumer.getCacheSize());
     consumer.create();
+    this.context = context;
+    operatorId = context.getId();
+    idempotentStorageManager.setup(context);
+    offsetStats.clear();
```
Comment (chandnisingh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractKafkaInputOperator.java]: the offsetStats is transient and will not have any state so why clear it?

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractKafkaInputOperator.java [hunk 2]
```diff
@@ -115,15 +161,91 @@ public void teardown()
...
+  {
+    try {
+      @SuppressWarnings("unchecked")
+      Map<KafkaPartition, KafkaPair<Long, Integer>> recoveredData = (Map<KafkaPartition, KafkaPair<Long, Integer>>) idempotentStorageManager.load(operatorId, windowId);
+      if (recoveredData == null) {
+        return;
+      }
 
+      System.out.println("In Replay State: " + recoveredData.size());
+
```
Comment (chandnisingh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractKafkaInputOperator.java]: Please remove System.out.println

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractKafkaInputOperator.java [hunk 3]
```diff
@@ -115,15 +161,91 @@ public void teardown()
...
+      Map<String, List<PartitionMetadata>> pms = KafkaMetadataUtil.getPartitionsForTopic(getConsumer().brokers, getConsumer().topic);
+      if (pms == null) {
+       return;
+      }
+
+      if(!(getConsumer() instanceof  SimpleKafkaConsumer))
+        return;
```
Comment (chandnisingh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractKafkaInputOperator.java]: I reckon you still have not updated IDE settings. Can you please do so.

Comment (chandnisingh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractKafkaInputOperator.java]: Also if you are setting NoopIdempotentStorageManager for any other kafka consumer, then the condition currentWindowId <= getLargestWindowId will not be true and replay() will not be called. 
So there is no need for above check

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractKafkaInputOperator.java [hunk 4]
```diff
@@ -165,12 +298,41 @@ public void deactivate()
...
+      {
+        currentWindowRecoveryState.put(consumerData.kafkaPart, new KafkaPair<Long, Integer>(consumerData.offSet, 1));
+      } else {
+        Pair<Long, Integer> second = currentWindowRecoveryState.get(consumerData.kafkaPart);
+        Integer noOfMessages = second.getSecond();
+        currentWindowRecoveryState.put(consumerData.kafkaPart, new KafkaPair<Long, Integer>(second.getFirst(), noOfMessages+1));
+      }
+    }
+    if(isReplayState)
+    {
```
Comment (chandnisingh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractKafkaInputOperator.java]: IsReplayState should never be true if NoopIdempotentStorageManager is set. So there should not be a need to check for the consumer type

**Comment by chaithu14** [comment]
For HighLevelKafkaConsumer, I have seen, the condition currentWindowId <= getLargestWindowId  is true. That's why I added those statements. The issue is creating the IdempotentStorageManager in constructor based on the consumer type. At this point, consumer type is always SimpleKafkaConsumer.

File: contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractKafkaInputOperator.java [hunk 5]
```diff
@@ -165,12 +289,39 @@ public void deactivate()
...
+      SimpleKafkaConsumer cons = (SimpleKafkaConsumer)getConsumer();
+      // Set the offset positions to the consumer
+      Map<KafkaPartition, Long> currentOffsets = new HashMap<KafkaPartition, Long>(cons.getCurrentOffsets());
+      // Increment the offsets
+      for (Map.Entry<KafkaPartition, Long> e: offsetStats.entrySet())
+        offsetStats.put(e.getKey(), e.getValue() + 1);
+
+      currentOffsets.putAll(offsetStats);
+      cons.resetOffset(currentOffsets);
+      cons.start();
```
Comment (chandnisingh) [inline contrib/src/main/java/com/datatorrent/contrib/kafka/AbstractKafkaInputOperator.java]: Should this block run just after the operator has finished replaying all the windows?
If that is the case then it should be place before the block above it.

**Comment by chandnisingh** [comment]
Since there were modifications to emitTuples() wrt to my earlier comment, can you please add test cases that will verify emitTuples once all the recovery windows are replayed

**Pull request closed without merging**